pennylane_lightning/core/src/simulators/lightning_qubit/catalyst/StateVectorLQubitDynamic.hpp

```diff
@@ -0,0 +1,387 @@
+// Copyright 2022-2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#pragma once
+
+#include <numeric>
+#include <utility>
+#include <vector>
+
+#include "BitUtil.hpp"
+#include "LinearAlgebra.hpp"
+#include "Util.hpp"
+
+#include "Error.hpp"
+
+#include <StateVectorLQubit.hpp>
+
+#include <iostream>
+
+namespace {
+using Pennylane::Util::AlignedAllocator;
+using Pennylane::Util::bestCPUMemoryModel;
+using Pennylane::Util::exp2;
+using Pennylane::Util::isPerfectPowerOf2;
+using Pennylane::Util::log2PerfectPower;
+using Pennylane::Util::ONE;
+using Pennylane::Util::squaredNorm;
+using Pennylane::Util::ZERO;
+} // namespace
+
+namespace Pennylane::LightningQubit {
+/**
+ * @brief State-vector dynamic class.
+ *
+ * This class allocates and deallocates qubits/wires dynamically,
+ * and defines all operations to manipulate the statevector data for
+ * quantum circuit simulation.
+ *
+ */
+template <class fp_t = double>
+class StateVectorLQubitDynamic
+    : public StateVectorLQubit<fp_t, StateVectorLQubitDynamic<fp_t>> {
+  public:
+    using PrecisionT = fp_t;
+    using ComplexT = std::complex<PrecisionT>;
+    using CFP_t = ComplexT;
+    using MemoryStorageT = Pennylane::Util::MemoryStorageLocation::Internal;
+
+  private:
+    using BaseType =
+        StateVectorLQubit<PrecisionT, StateVectorLQubitDynamic<PrecisionT>>;
+    std::vector<ComplexT, AlignedAllocator<ComplexT>> data_;
+
+    static constexpr PrecisionT epsilon_ =
+        std::numeric_limits<PrecisionT>::epsilon() * 100;
+
+    template <class IIter, class OIter>
+    inline OIter _move_data_elements(IIter first, size_t distance,
+                                     OIter second) {
+        *second++ = std::move(*first);
+        for (size_t i = 1; i < distance; i++) {
+            *second++ = std::move(*++first);
```

Comment:
Do we have a test that hits this, or is codecov just being a pain?

Reply:
I guess we don't have a test for this case

Reply:
No, we don't have explicit tests for these private helper methods! But they used to be tested in the device C++ unit tests.

Reply:
Fixed